dbaccess/block.go

```diff
@@ -0,0 +1,257 @@
+	if err != nil {
+		return nil, err
+	}
+
+	query := db.Where(&dbmodels.Block{BlockHash: blockHash})
+	query = preloadColumns(query, preloadedColumns)
+
+	block := &dbmodels.Block{}
+	dbResult := query.First(block)
+
+	dbErrors := dbResult.GetErrors()
+	if httpserverutils.IsDBRecordNotFoundError(dbErrors) {
+		return nil, nil
+	}
+	if httpserverutils.HasDBError(dbErrors) {
+		return nil, httpserverutils.NewErrorFromDBErrors("Some errors were encountered when loading block from database:",
+			dbResult.GetErrors())
+	}
+
+	return block, nil
+}
+
+// BlocksByHashes retreives a list of blocks with the corresponding `hashes`
+func BlocksByHashes(ctx Context, hashes []string, preloadedColumns ...string) ([]*dbmodels.Block, error) {
+	db, err := ctx.db()
+	if err != nil {
+		return nil, err
+	}
+
+	query := db.Where("block_hash in (?)", hashes)
+	query = preloadColumns(query, preloadedColumns)
+
+	blocks := []*dbmodels.Block{}
+	dbResult := query.Find(&blocks)
+
+	dbErrors := dbResult.GetErrors()
+	if httpserverutils.HasDBError(dbErrors) {
+		return nil, httpserverutils.NewErrorFromDBErrors("Some errors were encountered when loading block from database:",
+			dbResult.GetErrors())
+	}
+
+	return blocks, nil
+}
+
+// Blocks retrieves from the database up to `limit` blocks in the requested `order`, skipping the first `skip` blocks
+// If preloadedColumns was provided - preloads the requested columns
+func Blocks(ctx Context, order Order, skip uint64, limit uint64, preloadedColumns ...string) ([]*dbmodels.Block, error) {
+	db, err := ctx.db()
+	if err != nil {
+		return nil, err
+	}
+
+	query := db.
+		Limit(limit).
+		Offset(skip)
+
+	if order != OrderUnknown {
+		query = query.Order(fmt.Sprintf("`id` %s", order))
+	}
+
+	query = preloadColumns(query, preloadedColumns)
+
+	blocks := []*dbmodels.Block{}
+	dbResult := query.Find(&blocks)
+
+	dbErrors := dbResult.GetErrors()
+	if httpserverutils.HasDBError(dbErrors) {
+		return nil, httpserverutils.NewErrorFromDBErrors("Some errors were encountered when loading blocks from the database:",
+			dbResult.GetErrors())
+	}
+
+	return blocks, nil
+}
+
+// SelectedTip fetches the selected tip from the database
+func SelectedTip(ctx Context) (*dbmodels.Block, error) {
+	db, err := ctx.db()
+	if err != nil {
+		return nil, err
+	}
+
+	block := &dbmodels.Block{}
+	dbResult := db.Order("blue_score DESC").
+		Where(&dbmodels.Block{IsChainBlock: true}).
+		First(block)
+
+	dbErrors := dbResult.GetErrors()
+	if httpserverutils.IsDBRecordNotFoundError(dbErrors) {
+		return nil, nil
+	}
+	if httpserverutils.HasDBError(dbErrors) {
+		return nil, httpserverutils.NewErrorFromDBErrors("Some errors were encountered when loading selected tip from the database:", dbErrors)
+	}
+
+	return block, nil
+}
+
+// BluestBlock fetches the block with the highest blue score from the database
+// Note this is not necessarily the same as SelectedTip(), since SelectedTip requires
+// the selected-parent-chain to be fully synced
```

Comment:
If the selected-parent-chain is fully synced(a.k.a. initial sync is over) - than SelectedTip and BluestBlock() will return the same thing.

BluestBlock() is specifically called during the initial sync, because selectedTip is still not known.

Theoretically we can remove .SelectedTip, but I'd rather have the distinction.